### PR TITLE
[Lang] Support formatted string with args in assert

### PR DESCRIPTION
**Pull request by k-ye**
This allows us to write 

```py
assert cond(), 'foo=%d bar=%f' % (foo, bar)
``` 

I think we can apply the same strategy to `print`, but let me focus on `assert` in this PR.

Related issue = #1434

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: python/taichi/lang/impl.py [hunk 1]
```diff
@@ -454,6 +454,15 @@ def fused_string(entries):
     taichi_lang_core.create_print(contentries)
 
 
+@taichi_scope
+def ti_assert(cond, msg, extra_args):
+    # Mostly a wrapper to help us convert from ti.Expr (defined in Python) to
+    # taichi_lang_core.Expr (defined in C++)
+    import taichi as ti
+    taichi_lang_core.create_assert_stmt(
+        ti.Expr(cond).ptr, msg, [ti.Expr(x).ptr for x in extra_args])
```
Comment (archibate) [inline python/taichi/lang/impl.py:463]: What if `extra_args` is `Vector`? Let's make use of `Vector.__ti_repr__` like currently `ti_print` does.

**Comment by archibate** [comment]
I'd suggest us to translate:
```py
assert cond, "foo=%f" % foo
```
into:
```py
if not cond:
  ti.ti_print("foo=%f" % foo)
  ti.core.insert_assert_failure_stmt()
```
for API consistence and maintainability. So that we could deprecate `AssertStmt`, and use `AssertFailureStmt` plus `IfStmt` instead.

**Review by archibate — CHANGES_REQUESTED** [review]
Thanks, this is a very useful feature. But as yuanming always reminds me, we should always discuss about the design, reach an agreement, before start to writing code. My opinion is to translate `assert` into `if` + `print` + `ti.assert_failure()`. What do you think?

File: tests/python/test_assert.py [hunk 2]
```diff
@@ -11,8 +11,13 @@ def test_assert_minimal():
     def func():
         assert 0
 
+    @ti.kernel
+    def func2():
+        assert False
+
     with pytest.raises(RuntimeError):
         func()
+        func2()
```
Comment (archibate) [inline tests/python/test_assert.py:20]:
```suggestion
    with pytest.raises(RuntimeError):
        func2()
```

File: tests/python/test_assert.py [hunk 3]
```diff
@@ -39,6 +44,28 @@ def func():
...
+
+    @ti.kernel
+    def assert_float():
+        y = 0.5
+        assert y < 0, 'y = %f' % y
+
+    with pytest.raises(RuntimeError, match=r'x\[10\] expect=0 got=42'):
+        assert_formatted()
+    with pytest.raises(RuntimeError, match=r'y = 0.5'):
+        assert_float()
```
Comment (archibate) [inline tests/python/test_assert.py:66]: Add an `assert_success` test too?

File: python/taichi/lang/impl.py (hunk 1 above)
Comment (k-ye) [reply python/taichi/lang/impl.py:463]: I see. Unfortunately, assert format specifier doesn't support such a fancy data structure yet, see https://github.com/taichi-dev/taichi/blob/c3f1a5d1899500eb08facca78fa1373e279cb692/taichi/program/program.cpp#L445-L454

It only has `%d` or `%f`..

**Comment by yuanming-hu** [comment]
> My opinion is to translate `assert` into `if` + `print` + `ti.assert_failure()`. What do you think?

Thanks for proposing. However, we actually want `AssertStmt` be a single and "atomic" statement instead of multiple statements. We need to write the assertion failure message to the result buffer, and using `print` will not do the same job.

File: tests/python/test_assert.py (hunk 3 above)
Comment (yuanming-hu) [reply tests/python/test_assert.py:66]: We actually already have
https://github.com/taichi-dev/taichi/blob/c3f1a5d1899500eb08facca78fa1373e279cb692/tests/python/test_assert.py#L42-L50

Here we should simply focus on formatting.

**Comment by k-ye** [comment]
> My opinion is to translate assert into if + print + ti.assert_failure(). What do you think?

Yeah, this seems cleaner. However, note that this is likely a larger change than expected:

1. `print` and `assert` handle args quite differently. Specifically, `print` doesn't support format specifiers yet, which would require the C++ side change.
2. We may also want to support printing to different std streams, so that assertion goes to `stderr`? (BTW, right now assertion messages are printed with `TI_ERROR`)

I'd propose to support this first just because users can start using it, then unify how print/assert handles their string arguments. `print` may be more complicated, because we need to support things like `print('foo=%d bar=%f' % (foo, bar), ' hello', 123)`.

**Comment by yuanming-hu** [comment]
> I'd propose to support this first just because users can start using it, then unify how print/assert handles their string arguments. `print` may be more complicated, because we need to support things like `print('foo=%d bar=%f' % (foo, bar), ' hello', 123)`.

Maybe we should also consider supporting f-strings in the future :-)

File: tests/python/test_assert.py (hunk 3 above)
Comment (k-ye) [reply tests/python/test_assert.py:67]: Hmm, anyway i flipped `x[0]` back to 0 and run `assert_formatted()` again for the success path.

**Comment by k-ye** [comment]
> Maybe we should also consider supporting f-strings in the future :-)

Ah yes, that sounds like a much bigger feature..

**Pull request merged**